src/connector/src/sink/remote.rs

```diff
@@ -241,80 +274,162 @@ impl LogSinker for RemoteLogSinker {
                 }
 
-                match item {
-                    LogStoreReadItem::StreamChunk { chunk, chunk_id } => {
-                        let offset = TruncateOffset::Chunk { epoch, chunk_id };
-                        if let Some(prev_offset) = &prev_offset {
-                            prev_offset.check_next_offset(offset)?;
+                log_reader.truncate(offset).await?;
+                Ok(())
+            }
+
+            let mut prev_offset: Option<TruncateOffset> = None;
+            // Push from back and pop from front
+            let mut sent_offset_queue: VecDeque<(TruncateOffset, Option<Instant>)> =
+                VecDeque::new();
+
+            loop {
+                let either_result: futures::future::Either<
+                    Option<SinkWriterStreamResponse>,
+                    anyhow::Result<(u64, LogStoreReadItem)>,
+                > = drop_either_future(
+                    select(pin!(response_rx.recv()), pin!(log_reader.next_item())).await,
+                );
+                match either_result {
+                    futures::future::Either::Left(opt) => {
+                        let response = opt.ok_or_else(|| anyhow!("end of response stream"))?;
+                        match response {
+                            SinkWriterStreamResponse {
+                                response:
+                                    Some(sink_writer_stream_response::Response::Batch(
+                                        sink_writer_stream_response::BatchWrittenResponse {
+                                            epoch,
+                                            batch_id,
+                                        },
+                                    )),
+                            } => {
+                                truncate_matched_offset(
+                                    &mut sent_offset_queue,
+                                    TruncateOffset::Chunk {
+                                        epoch,
+                                        chunk_id: batch_id as _,
+                                    },
+                                    &mut log_reader,
+                                    &sink_metrics,
+                                )
+                                .await?;
+                            }
+                            SinkWriterStreamResponse {
+                                response:
+                                    Some(sink_writer_stream_response::Response::Commit(
+                                        sink_writer_stream_response::CommitResponse {
+                                            epoch,
+                                            metadata,
+                                        },
+                                    )),
+                            } => {
+                                if let Some(metadata) = metadata {
+                                    warn!("get unexpected non-empty metadata: {:?}", metadata);
```

Comment:
Is this really unexpected? I saw that java iceberg sink will return non-empty metadata.

Reply:
iceberg sink uses `CoordinatedRemoteSinkWriter` and will not be affected by this PR. Sink that use this `RemoteLogSinker` assume that they are not coordinated and should not return non-empty metadata.